### PR TITLE
Add metrics integration to istio charms

### PR DESCRIPTION
**Pull request by IbraAoad**
## Issue
The Istio beacon exposes metrics through a separate workload it deploys:

- **`beacon-waypoint`**: Accessible at `istio-beacon-k8s-{self.model.name}-waypoint.{self.model.name}.svc:15020`, on port `15020` at `/stats/prometheus`.

The main issue with integrating via the beacon's FQDN is that Kubernetes will enforce load balancing when the beacon is scaled, resulting in Prometheus receiving inconsistent metrics.

## Solution
This PR adds metrics integration through the [prometheus_scrape](https://charmhub.io/prometheus-k8s/libraries/prometheus_scrape) interface, utilizing the newly created [metrics-k8s-proxy](https://github.com/canonical/metrics-k8s-proxy) to aggregate these metrics into Prometheus via a unified endpoint.

## Testing Instructions

1. juju deploy istio-beacon-k8s --trust
2. juju deploy prometheus-k8s --trust
3. juju relate istio-k8s istio-beacon-k8s
4. Explore Prometheus to validate that the metrics are present. Additionally, check for the `up` metric for each the waypoint pod.

### CODE REVIEW

**Comment by IbraAoad** [comment]
Implementation is ready for review

**TODO**: Fix proxy image location + integration tests once https://github.com/canonical/metrics-k8s-proxy/pull/1 and https://github.com/canonical/metrics-proxy-rock/pull/1 are merged

File: charmcraft.yaml [hunk 1]
```diff
@@ -61,4 +67,10 @@ config:
...
         Add this charm's model to the service mesh. 
-        All charms in this model will automatically be added to the mesh.
\ No newline at end of file
+        All charms in this model will automatically be added to the mesh.
+
+resources:
+  metrics-proxy-image:
+    type: oci-image
+    description: Metrics Proxy OCI image
+    upstream-source: localhost:32000/metrics-proxy:0.1
```
Comment (dstathis) [inline charmcraft.yaml:76]: This should be the upstream image rather than localhost.

File: src/charm.py [hunk 2]
```diff
@@ -32,6 +34,8 @@
 WAYPOINT_RESOURCE_TYPES = {RESOURCE_TYPES["Gateway"]}
 WAYPOINT_LABEL = "istio-waypoint"
 
+METRICS_LABELS = {"charms.canonical.com/istio.io.waypoint.metrics": "aggregated"}
```
Comment (dstathis) [inline src/charm.py:37]: The label should include model-uuid and app-name. This is necessary (unlike for istio-k8s where it is just nice to have). With this label we will scrape the metrics from every beacon charm right?

File: src/charm.py [hunk 3]
```diff
@@ -42,13 +46,50 @@ def __init__(self, *args):
         self._lightkube_field_manager: str = self.app.name
         self._lightkube_client = None
         self._managed_labels = f"{self.app.name}-{self.model.name}"
+        self._metrics_labels = ",".join(
+            [f"{key}={value}" for key, value in METRICS_LABELS.items()]
+        )
+        # Configure Observability
+        self._scraping = MetricsEndpointProvider(
+            self,
+            relation_name="metrics-endpoint",
```
Comment (dstathis) [inline src/charm.py:55]: No need to specify the default relation_name value.

File: src/charm.py [hunk 4]
```diff
@@ -42,13 +46,50 @@ def __init__(self, *args):
...
+            jobs=[{"static_configs": [{"targets": ["*:15090"]}]}],
+        )
+
         self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.remove, self._on_remove)
         self._mesh = ServiceMeshProvider(self, labels=self.mesh_labels())
 
         self.framework.observe(self.on["service-mesh"].relation_changed, self.on_mesh_changed)
         self.framework.observe(self.on["service-mesh"].relation_broken, self.on_mesh_broken)
```
Comment (dstathis) [inline src/charm.py:64]: We need to observe pebble_ready. See [istio-k8s comment](https://github.com/canonical/istio-k8s-operator/pull/20#discussion_r1799962891).

**Review by dstathis — APPROVED** [review]

**Pull request merged**